src/status_im2/contexts/add_new_contact/events_test.cljs

```diff
@@ -2,52 +2,57 @@
-   :networks/networks        {"mainnet_rpc"
-                              {:id     "mainnet_rpc"
-                               :config {:NetworkId 1}}}})
-
-(defn search-db
-  [input]
-  {:contacts/new-identity {:input      input
-                           :public-key nil
-                           :ens-name   nil
-                           :state      :searching
-                           :error      nil}})
+(def ukey
+  "0x045596a7ff87da36860a84b0908191ce60a504afc94aac93c1abd774f182967ce694f1bf2d8773cd59f4dd0863e951f9b7f7351c5516291a0fceb73f8c392a0e88")
+(def ckey "zQ3shWj4WaBdf2zYKCkXe6PHxDxNTzZyid1i75879Ue9cX9gA")
+(def ens "esep")
+
+(deftest identify-empty-string
+  (is (= (core/identify-type "")
+         {:input    ""
+          :id       nil
+          :type     :empty
+          :ens-name nil})))
 
 (deftest search-empty-string
-  (let [input    ""
-        expected {:db init-db}
-        actual   (core/set-new-identity {:db init-db} input)]
-    (is (= actual expected))))
-
-(deftest search-ens
-  (let [input    "esep"
-        clean    (fn [db]
-                   (-> db
-                       (assoc-in [:contacts/resolve-public-key-from-ens-name :on-success] nil)
-                       (assoc-in [:contacts/resolve-public-key-from-ens-name :on-error] nil)))
-        expected {:db                                        (merge init-db (search-db input))
-                  :contacts/resolve-public-key-from-ens-name
-                  {:chain-id   1
-                   :ens-name   (str input ".stateofus.eth")
-                   :on-success nil
-                   :on-error   nil}}
-        actual   (core/set-new-identity {:db init-db} input)]
-    (is (= (clean actual) expected))))
-
-(deftest search-compressed-key
-  (let [input    "zQ3shWj4WaBdf2zYKCkXe6PHxDxNTzZyid1i75879Ue9cX9gA"
-        clean    (fn [db]
-                   (-> db
-                       (assoc-in [:contacts/decompress-public-key :on-success] nil)
-                       (assoc-in [:contacts/decompress-public-key :on-error] nil)))
-        expected {:db                             (merge init-db (search-db input))
-                  :contacts/decompress-public-key
-                  {:public-key input
-                   :on-success nil
-                   :on-error   nil}}
-        actual   (core/set-new-identity {:db init-db} input)]
-    (is (= (clean actual) expected))))
+  (is (= (core/set-new-identity {:db {:contacts/new-identity :foo}} "")
+         {:db {}})))
+
+(deftest identify-uncompressed-key
+  (is (= (core/identify-type ukey)
+         {:input    ukey
+          :id       ukey
+          :type     :public-key
+          :ens-name nil})))
+
+(deftest search-uncompressed-key
+  (is (= (core/set-new-identity {:db {}} ukey)
+         {:db {:contacts/new-identity
+               {:input      ukey
+                :public-key ukey
+                :state      :error
+                :error      :uncompressed-key}}})))
+
+(deftest identify-ens
```

Comment:
> Related to that, I see the tests for `identify-type` are too "optimistic" (happy path only). For instance, the function `identify-type` incorrectly identifies anything that's not `empty|public-key|compressed-key` as an `:ens-name`. Having an easier way to add more test cases could've helped you catch this kind of problem.

Thanks for looking closely at that! However, I'm not sure if I agree that it's incorrect or a problem. What input do you think is problematic?

Reply:
> Going further, I think the `ens` validation is flawed. I think we should follow the specification https://eips.ethereum.org/EIPS/eip-137#name-syntax and never fallback to `ens`. The event handler that sets the identity should be more restrictive, otherwise the app state will be incorrect and other problems will follow.

Ah, ens... I did not put in more extensive validation because:
- This feature belongs to another PR: #14478
- ens names seem to be quite permissive (I'll take another look at it, thanks for the link!)
- status-go will attempt to resolve it anyway

Reply:
> The regex for the URL should have an anchor `^` (beginning of line) to avoid matching strings that don't start with the HTTP protocol. Node also offers a native URL parser that would allow us to extract the ID in a more future proof way (for instance if the the invite URL had query params), but I find it unnecessary for this case. Just for the sake of sharing knowledge and because I was curious to refresh my memory:
> 
> ```clojure
> (defn invite-url->id
>   [invite-url]
>   (when-let [url (and (re-find #"^https?://join.status.im/u/.+" invite-url)
>                       (try (new js/URL invite-url)
>                            (catch js/Error _ nil)))]
>     (let [pathname (.-pathname url)
>           id       (nth (string/split pathname "/") 2)]
>       id)))
> 
> (comment
>   (assert (= "some-id" (invite-url->id "https://join.status.im/u/some-id?a=1&b=2"))))
> ```

I had included the caret earlier, but took it out after I started trimming the string. I'll add it back in, np!

Thanks for sharing the code snippet, interesting!

Reply:
Again sharing in case someone is not aware, the `are` macro is quite flexible and you can design the assertions to eliminate duplication because when using the `are` macro, the whole point is to easily describe input + expected.

This passes in the test suite:

```clojure
(deftest identify-type-test
  (are [args] (= (assoc (:expected args) :input (:input args))
                 (core/identify-type (:input args)))
   {:input    ukey
    :expected {:id       ukey
               :type     :public-key
               :ens-name nil}}
   (let [ens-name (str ens ".stateofus.eth")]
     {:input    ens
      :expected {:id       ens
                 :type     :ens-name
                 :ens-name ens-name}})
   (let [input (str "https://join.status.im/u/" ckey)]
     {:input    input
      :expected {:id       ckey
                 :type     :compressed-key
                 :ens-name nil}})))
```

Reply:
Thanks Icaro, I got `are` working, here's what I'll probably commit. (defs at the top for the other test to use)
``` clojure
(def ukey
  "0x045596a7ff87da36860a84b0908191ce60a504afc94aac93c1abd774f182967ce694f1bf2d8773cd59f4dd0863e951f9b7f7351c5516291a0fceb73f8c392a0e88")
(def ckey "zQ3shWj4WaBdf2zYKCkXe6PHxDxNTzZyid1i75879Ue9cX9gA")
(def ens "esep")
(def link-ckey (str "https://join.status.im/u/" ckey))

(deftest identify-type-test
  (are [input expected] (= (core/identify-type input) expected)
    "" {:input    ""
        :id       nil
        :type     :empty
        :ens-name nil}    

    ukey {:input    ukey
          :id       ukey
          :type     :public-key
          :ens-name nil}

    ens {:input    ens
         :id       ens
         :type     :ens-name
         :ens-name (str ens ".stateofus.eth")}

    ckey {:input    ckey
          :id       ckey
          :type     :compressed-key
          :ens-name nil}

    link-ckey {:input    link-ckey
               :id       ckey
               :type     :compressed-key
               :ens-name nil}))
```